src/Misc/MiddleWare.cpp

```diff
@@ -298,14 +298,49 @@ struct NonRtObjStore
         return objmap[loc];
     }
 
-    void handleOscil(const char *msg, rtosc::RtData &d) {
-        string obj_rl(d.message, msg);
+    //! try to dispatch a message at the OscilGen non-RT-ports
+    void handleOscilADnote(const char *msg, rtosc::RtData &d, bool voiceOscil) {
+        const string obj_rl(d.message, msg);
         void *osc = get(obj_rl);
         if(osc)
         {
             strcpy(d.loc, obj_rl.c_str());
             d.obj = osc;
             OscilGen::non_realtime_ports.dispatch(msg, d);
+            if(d.matches == 0 /* || d.forwarded -> check should be useless, can't be forwarded */)
+            {
+                // nothing to re-compute
+                // this will probably never be the case when OscilGen::get will be non-RT
+            }
+            else
+            {
+                if(voiceOscil)
+                {
+                    OscilGen* oscilGen = static_cast<OscilGen*>(d.obj);
+                    WaveTable wt = oscilGen->calculateWaveTable();
+                    // TODO: how to send wt back to ADnoteVoice?
+                    //       maybe via OSC, like this?
+#if 0
```

Comment:
Looks like you have the general idea of:
1) get the path to the oscillator
2) go up one level
3) turn that into a method which will be associated with ADnoteVoiceParam

it's missing
4) get pointers to memory representing either buffers or pointers to structs
5) send them with a (I think) d.forward()
6) Receive them on the backend and insert those values into the realtime side wavetable.

Reply:
`d.forward()`? That's a nice trick to get rid of creating another `RtData` object! We could use it to send a message to `ADvoiceParameters`, telling it to insert the new buffers/pointer to structs. I could modify the `d.loc` by writing a `0` into the string to cut the `OscilSmp/`, but do you have an idea how to modify the `msg`? It's right now `const`.

In case we use a thread for the wavetable computation, I think we can't use `d.forward()` anymore.

Reply:
One way to modify the strings would be to handle this case separately, directly before inserting into `uToB` (line 2101). I could create a temporary string there for this special case. That's the only idea I have for now.

Reply:
I mentioned this in IRC, but I was misremembering the API. d.forward() should just be "I'm not handling this, the other threading context will" d.chain() should be "send the message I created onto the next thread's handler

Reply:
As per modifying the message, you don't. Treat the message as a constant and create a new message to send onwards. I think d.chain() has the typesignature of d.chain(const char *path, const char *args, ...) You should not need to explicitly reference uToB in that callback (and it's best if you don't as it's easier to instrument what's being sent if the RtData methods are used).

Reply:
Solved using `d.chain` as demanded.